hdl/BitSampling.bsv

```diff
@@ -12,56 +12,89 @@ import DReg::*;
 import GetPut::*;
 import Vector::*;
 
-
+//
+// A `BitSampler` can be used to implement sampling primitives which recover
+// received bits from an incoming signal. The number of samples per bit is given
+// by the `bit_period`.
+//
 interface BitSampler #(numeric type bit_period);
     interface Put#(Bit#(1)) in;
     interface Get#(Bit#(1)) out;
 endinterface
 
+//
+// `mkBitSampler` is an asynchronous implementation of the `BitSampler`
+// interface. It assumes a source producing a bit signal with a constant given
+// bit period.
+//
+// Without any bit edges in the signal the sampler will provide a sampled bit
+// every `bit_period` iterations. Any time an edge is detected in the signal the
+// sample point is re-aligned in an attempt to sample the next bit near the
+// center of the bit period. This allows the sampler to track the signal source
+// despite some amount of clock jitter and sampling aliasing. This works
+// particularly well for signals with short run lengths but can also be applied
+// to signals witn an asynchronous encoding.
+//
+// It is theoretically possible to use a bit period of three samples on external
+// signals if the clocks in both systems have sufficient jitter performance.
+// Samplers with a bit period of five and more samples have been succesfully
+// demonstrated using real hardware, with both synchronous (8B10B encoded)
+// signals and asynchronous encoded (UART) signals. A bit period of two can only
+// be used between systems with an alias free channel.
+//
+// Note that due to the sampler re-aligning on every bit edge it is not suited
+// for use with signals with uneven bit periods, such as a PWM or I^2C signal.
+//
 module mkBitSampler (BitSampler#(bit_period))
         provisos (
             // The bit period should last two or more samples.
             Add#(2, a__, bit_period),
             // Ceiling of bit_period / 2. This works correct for odd count
             // periods.
             Div#(bit_period, 2, half_bit_period));
+
     ConfigReg#(Vector#(2, Bit#(1))) samples <- mkConfigRegU();
     Reg#(UInt#(TLog#(bit_period))) count <- mkRegU();
     Reg#(Bool) sample_valid <- mkDReg(False);
 
     interface Put in;
         method Action put(Bit#(1) b);
+            // Sampler events.
             let bit_edge = (samples[1] != samples[0]);
-            let do_sample = (count == 0);
+            let count_zero = (count == 0);
 
+            // Shift in the next sample.
             samples <= shiftInAtN(samples, b);
```

Comment:
I feel like the word sample here is a bit overloaded, took me a bit to realize that while we're sampling every Put (and shifting the samples vec), this block is choosing when to say it's valid and show it downstream which is what the higher-level design is considering a "sample."

That said, I'm not sure I have much of a better suggestion for naming. sample[1] is really only used for your edge detector.
 My go-to here would have been to always store last_put or something, but look at b combinatorially and only snap it to a register when it was valid, thus the output register would always be valid even if you still wanted to gate the out method with a valid flag to preserve the interface.  That would change the names and flow but I don't really know if it's better.  #helpfulComments

Reply:
Maybe where I come down is that this module isn't really a Sampler so much as also sample aligner?

Reply:
Yes, I modeled this as a shift register/"sliding window" over the incoming signal, with the assumption that downstream logic can accept a sample the very next cycle the bit center has been determined. If this can not be met a FIFO (or a simple register) after the `out` interface should be provided by downstream logic. I've documented this assumption.

I've tried to overload the terms a bit less and improve the naming.